adapters/akka-http/src/main/scala/caliban/AkkaHttpAdapter.scala

```diff
@@ -121,16 +126,39 @@ trait AkkaHttpAdapter {
 
     get {
       extractUpgradeToWebSocket { upgrade =>
-        val (queue, source) = Source.queue[Message](0, OverflowStrategy.fail).preMaterialize()
-        val subscriptions   = runtime.unsafeRun(Ref.make(Map.empty[String, Fiber[Throwable, Unit]]))
+        val (queue: SourceQueueWithComplete[Message], source) =
+          Source.queue[Message](0, OverflowStrategy.fail).preMaterialize()
+        val subscriptions = runtime.unsafeRun(Ref.make(Map.empty[String, Fiber[Throwable, Unit]]))
         val sink = Sink.foreach[Message] {
           case TextMessage.Strict(text) =>
             val io = for {
               msg     <- Task.fromEither(json.parseWSMessage(text))
               msgType = msg.messageType
               _ <- IO.whenCase(msgType) {
                     case "connection_init" =>
-                      IO.fromFuture(_ => queue.offer(TextMessage("""{"type":"connection_ack"}""")))
+                      val ack: ZIO[Clock with Random, Throwable, Any] =
+                        IO.fromFuture(_ => queue.offer(TextMessage("""{"type":"connection_ack"}""")))
+
+                      val withKeepAlive = keepAliveTime.fold(ack) { time =>
+                        ack
+                          .flatMap(_ =>
+                            IO.fromFuture(_ =>
+                                queue
+                                  .offer(
+                                    TextMessage("""{"type":"ka"}""")
+                                  )
+                              )
+                              .repeat(Schedule.spaced(time).jittered)
+                              .forkDaemon
```

Comment:
This fiber will go on forever, you need to capture it and interrupt it when the connection is stopped.

Reply:
Please take a look at my change, something like that?